sematic/future_context.py

```diff
@@ -1,35 +1,39 @@
 # Standard Library
 from contextlib import contextmanager
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from importlib import import_module
-from typing import Optional, Type
+from typing import Any, List, Optional, Type
 
-# Sematic
-from sematic.resolver import Resolver
+# To avoid a circular dependency of:
+# future_context -> resolver -> future -> future_context
+Resolver = Any
```

Comment:
Yes, hence my suggestion to decouple future from context (i.e. decouple from external resources). Context is naturally aware of futures, but ideally futures are not aware of context. This would remove this circular dependency.

Reply:
The question is, how do you make it so that futures created inside a with block get the resource associated with them while others do not?

```
with SomeResource as r:
    x = do_something(r)
return do_something_else(x)
```

This *requires* that from either the future or the calculator, you have *some way* to get a handle on the resources (which are in the context). You might think having the calculator be aware of the need for this coupling would make things easier, but then the calculator needs some way to communicate to the resolver instance about the coupling. You can do this by putting the association back in the future, but then you haven't really gained anything. Alternatively, you can give the calculator a reference to the resolver, but that isn't really any better because then you have the calculator depending on resolvers, while the resolvers need to also be aware of calculators.

Here's the most indirect way I can think of (indirect as in it doesn't create dependencies of future on context or context on resolver):
- have the futures (or calculators) have class-attribute-level "creation callbacks" that they invoke any time they are created
- have the context manager register a creation callback on `__enter__` and de-register on `__exit__` (i.e. `Future.register_callback(context.on_future_created)`)
- give the context "future resource connection" callbacks that can be registered
- have the resolvers register themselves to get called on "future resource connection"
- in the context's `on_future_created`, it goes through its `future_resource_connection` callbacks and calls them with the created future as well as whatever resources were active at the time

this seems like a lot of indirection that will make the code harder to follow and understand, but it will remove the circular dependencies and arguably increase encapsulation.